// @mysten/walrus@0.0.19/dist/cjs/contracts/deps/0x0000000000000000000000000000000000000000000000000000000000000002/authenticator_state.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var authenticator_state_exports = {};
__export(authenticator_state_exports, {
  ActiveJwk: () => ActiveJwk,
  AuthenticatorState: () => AuthenticatorState,
  AuthenticatorStateInner: () => AuthenticatorStateInner,
  JWK: () => JWK,
  JwkId: () => JwkId
});
module.exports = __toCommonJS(authenticator_state_exports);
var import_bcs = require("@mysten/sui/bcs");
var object = __toESM(require("./object.js"));
function AuthenticatorState() {
  return import_bcs.bcs.struct("AuthenticatorState", {
    id: object.UID(),
    version: import_bcs.bcs.u64()
  });
}
function AuthenticatorStateInner() {
  return import_bcs.bcs.struct("AuthenticatorStateInner", {
    version: import_bcs.bcs.u64(),
    active_jwks: import_bcs.bcs.vector(ActiveJwk())
  });
}
function JWK() {
  return import_bcs.bcs.struct("JWK", {
    kty: import_bcs.bcs.string(),
    e: import_bcs.bcs.string(),
    n: import_bcs.bcs.string(),
    alg: import_bcs.bcs.string()
  });
}
function JwkId() {
  return import_bcs.bcs.struct("JwkId", {
    iss: import_bcs.bcs.string(),
    kid: import_bcs.bcs.string()
  });
}
function ActiveJwk() {
  return import_bcs.bcs.struct("ActiveJwk", {
    jwk_id: JwkId(),
    jwk: JWK(),
    epoch: import_bcs.bcs.u64()
  });
}
