// @mysten/walrus@0.0.19/dist/cjs/contracts/deps/0x0000000000000000000000000000000000000000000000000000000000000002/bls12381.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var bls12381_exports = {};
__export(bls12381_exports, {
  G1: () => G1,
  G2: () => G2,
  GT: () => GT,
  Scalar: () => Scalar,
  UncompressedG1: () => UncompressedG1
});
module.exports = __toCommonJS(bls12381_exports);
var import_bcs = require("@mysten/sui/bcs");
function Scalar() {
  return import_bcs.bcs.struct("Scalar", {
    dummy_field: import_bcs.bcs.bool()
  });
}
function G1() {
  return import_bcs.bcs.struct("G1", {
    dummy_field: import_bcs.bcs.bool()
  });
}
function G2() {
  return import_bcs.bcs.struct("G2", {
    dummy_field: import_bcs.bcs.bool()
  });
}
function GT() {
  return import_bcs.bcs.struct("GT", {
    dummy_field: import_bcs.bcs.bool()
  });
}
function UncompressedG1() {
  return import_bcs.bcs.struct("UncompressedG1", {
    dummy_field: import_bcs.bcs.bool()
  });
}
