// @mysten/walrus@0.0.19/dist/cjs/contracts/deps/0x0000000000000000000000000000000000000000000000000000000000000002/deny_list.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var deny_list_exports = {};
__export(deny_list_exports, {
  AddressKey: () => AddressKey,
  ConfigKey: () => ConfigKey,
  ConfigWriteCap: () => ConfigWriteCap,
  DenyList: () => DenyList,
  GlobalPauseKey: () => GlobalPauseKey,
  PerTypeConfigCreated: () => PerTypeConfigCreated,
  PerTypeList: () => PerTypeList
});
module.exports = __toCommonJS(deny_list_exports);
var import_bcs = require("@mysten/sui/bcs");
var bag = __toESM(require("./bag.js"));
var object = __toESM(require("./object.js"));
var table = __toESM(require("./table.js"));
function DenyList() {
  return import_bcs.bcs.struct("DenyList", {
    id: object.UID(),
    lists: bag.Bag()
  });
}
function ConfigWriteCap() {
  return import_bcs.bcs.struct("ConfigWriteCap", {
    dummy_field: import_bcs.bcs.bool()
  });
}
function ConfigKey() {
  return import_bcs.bcs.struct("ConfigKey", {
    per_type_index: import_bcs.bcs.u64(),
    per_type_key: import_bcs.bcs.vector(import_bcs.bcs.u8())
  });
}
function AddressKey() {
  return import_bcs.bcs.struct("AddressKey", {
    pos0: import_bcs.bcs.Address
  });
}
function GlobalPauseKey() {
  return import_bcs.bcs.struct("GlobalPauseKey", {
    dummy_field: import_bcs.bcs.bool()
  });
}
function PerTypeConfigCreated() {
  return import_bcs.bcs.struct("PerTypeConfigCreated", {
    key: ConfigKey(),
    config_id: import_bcs.bcs.Address
  });
}
function PerTypeList() {
  return import_bcs.bcs.struct("PerTypeList", {
    id: object.UID(),
    denied_count: table.Table(),
    denied_addresses: table.Table()
  });
}
