// @mysten/walrus@0.0.19/dist/cjs/contracts/deps/0x0000000000000000000000000000000000000000000000000000000000000002/kiosk.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var kiosk_exports = {};
__export(kiosk_exports, {
  Borrow: () => Borrow,
  Item: () => Item,
  ItemDelisted: () => ItemDelisted,
  ItemListed: () => ItemListed,
  ItemPurchased: () => ItemPurchased,
  Kiosk: () => Kiosk,
  KioskOwnerCap: () => KioskOwnerCap,
  Listing: () => Listing,
  Lock: () => Lock,
  PurchaseCap: () => PurchaseCap
});
module.exports = __toCommonJS(kiosk_exports);
var import_bcs = require("@mysten/sui/bcs");
var balance = __toESM(require("./balance.js"));
var object = __toESM(require("./object.js"));
function Kiosk() {
  return import_bcs.bcs.struct("Kiosk", {
    id: object.UID(),
    profits: balance.Balance(),
    owner: import_bcs.bcs.Address,
    item_count: import_bcs.bcs.u32(),
    allow_extensions: import_bcs.bcs.bool()
  });
}
function KioskOwnerCap() {
  return import_bcs.bcs.struct("KioskOwnerCap", {
    id: object.UID(),
    for: import_bcs.bcs.Address
  });
}
function PurchaseCap() {
  return import_bcs.bcs.struct("PurchaseCap", {
    id: object.UID(),
    kiosk_id: import_bcs.bcs.Address,
    item_id: import_bcs.bcs.Address,
    min_price: import_bcs.bcs.u64()
  });
}
function Borrow() {
  return import_bcs.bcs.struct("Borrow", {
    kiosk_id: import_bcs.bcs.Address,
    item_id: import_bcs.bcs.Address
  });
}
function Item() {
  return import_bcs.bcs.struct("Item", {
    id: import_bcs.bcs.Address
  });
}
function Listing() {
  return import_bcs.bcs.struct("Listing", {
    id: import_bcs.bcs.Address,
    is_exclusive: import_bcs.bcs.bool()
  });
}
function Lock() {
  return import_bcs.bcs.struct("Lock", {
    id: import_bcs.bcs.Address
  });
}
function ItemListed() {
  return import_bcs.bcs.struct("ItemListed", {
    kiosk: import_bcs.bcs.Address,
    id: import_bcs.bcs.Address,
    price: import_bcs.bcs.u64()
  });
}
function ItemPurchased() {
  return import_bcs.bcs.struct("ItemPurchased", {
    kiosk: import_bcs.bcs.Address,
    id: import_bcs.bcs.Address,
    price: import_bcs.bcs.u64()
  });
}
function ItemDelisted() {
  return import_bcs.bcs.struct("ItemDelisted", {
    kiosk: import_bcs.bcs.Address,
    id: import_bcs.bcs.Address
  });
}
