// @mysten/walrus@0.0.19/dist/cjs/contracts/deps/0x0000000000000000000000000000000000000000000000000000000000000002/linked_table.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var linked_table_exports = {};
__export(linked_table_exports, {
  LinkedTable: () => LinkedTable,
  Node: () => Node
});
module.exports = __toCommonJS(linked_table_exports);
var import_bcs = require("@mysten/sui/bcs");
var object = __toESM(require("./object.js"));
function LinkedTable(...typeParameters) {
  return import_bcs.bcs.struct("LinkedTable", {
    id: object.UID(),
    size: import_bcs.bcs.u64(),
    head: import_bcs.bcs.option(typeParameters[0]),
    tail: import_bcs.bcs.option(typeParameters[0])
  });
}
function Node(...typeParameters) {
  return import_bcs.bcs.struct("Node", {
    prev: import_bcs.bcs.option(typeParameters[0]),
    next: import_bcs.bcs.option(typeParameters[0]),
    value: typeParameters[1]
  });
}
