// @mysten/walrus@0.0.19/dist/cjs/contracts/deps/0x0000000000000000000000000000000000000000000000000000000000000002/package.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var package_exports = {};
__export(package_exports, {
  Publisher: () => Publisher,
  UpgradeCap: () => UpgradeCap,
  UpgradeReceipt: () => UpgradeReceipt,
  UpgradeTicket: () => UpgradeTicket
});
module.exports = __toCommonJS(package_exports);
var import_bcs = require("@mysten/sui/bcs");
var object = __toESM(require("./object.js"));
function Publisher() {
  return import_bcs.bcs.struct("Publisher", {
    id: object.UID(),
    package: import_bcs.bcs.string(),
    module_name: import_bcs.bcs.string()
  });
}
function UpgradeCap() {
  return import_bcs.bcs.struct("UpgradeCap", {
    id: object.UID(),
    package: import_bcs.bcs.Address,
    version: import_bcs.bcs.u64(),
    policy: import_bcs.bcs.u8()
  });
}
function UpgradeTicket() {
  return import_bcs.bcs.struct("UpgradeTicket", {
    cap: import_bcs.bcs.Address,
    package: import_bcs.bcs.Address,
    policy: import_bcs.bcs.u8(),
    digest: import_bcs.bcs.vector(import_bcs.bcs.u8())
  });
}
function UpgradeReceipt() {
  return import_bcs.bcs.struct("UpgradeReceipt", {
    cap: import_bcs.bcs.Address,
    package: import_bcs.bcs.Address
  });
}
