// @mysten/walrus@0.0.19/dist/cjs/contracts/deps/0x0000000000000000000000000000000000000000000000000000000000000002/token.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var token_exports = {};
__export(token_exports, {
  ActionRequest: () => ActionRequest,
  RuleKey: () => RuleKey,
  Token: () => Token,
  TokenPolicy: () => TokenPolicy,
  TokenPolicyCap: () => TokenPolicyCap,
  TokenPolicyCreated: () => TokenPolicyCreated
});
module.exports = __toCommonJS(token_exports);
var import_bcs = require("@mysten/sui/bcs");
var type_name = __toESM(require("../0x0000000000000000000000000000000000000000000000000000000000000001/type_name.js"));
var balance = __toESM(require("./balance.js"));
var object = __toESM(require("./object.js"));
var vec_map = __toESM(require("./vec_map.js"));
var vec_set = __toESM(require("./vec_set.js"));
function Token() {
  return import_bcs.bcs.struct("Token", {
    id: object.UID(),
    balance: balance.Balance()
  });
}
function TokenPolicyCap() {
  return import_bcs.bcs.struct("TokenPolicyCap", {
    id: object.UID(),
    for: import_bcs.bcs.Address
  });
}
function TokenPolicy() {
  return import_bcs.bcs.struct("TokenPolicy", {
    id: object.UID(),
    spent_balance: balance.Balance(),
    rules: vec_map.VecMap(import_bcs.bcs.string(), vec_set.VecSet(type_name.TypeName()))
  });
}
function ActionRequest() {
  return import_bcs.bcs.struct("ActionRequest", {
    name: import_bcs.bcs.string(),
    amount: import_bcs.bcs.u64(),
    sender: import_bcs.bcs.Address,
    recipient: import_bcs.bcs.option(import_bcs.bcs.Address),
    spent_balance: import_bcs.bcs.option(balance.Balance()),
    approvals: vec_set.VecSet(type_name.TypeName())
  });
}
function RuleKey() {
  return import_bcs.bcs.struct("RuleKey", {
    is_protected: import_bcs.bcs.bool()
  });
}
function TokenPolicyCreated() {
  return import_bcs.bcs.struct("TokenPolicyCreated", {
    id: import_bcs.bcs.Address,
    is_mutable: import_bcs.bcs.bool()
  });
}
