// @mysten/walrus@0.0.19/dist/cjs/contracts/deps/0x0000000000000000000000000000000000000000000000000000000000000002/transfer_policy.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var transfer_policy_exports = {};
__export(transfer_policy_exports, {
  RuleKey: () => RuleKey,
  TransferPolicy: () => TransferPolicy,
  TransferPolicyCap: () => TransferPolicyCap,
  TransferPolicyCreated: () => TransferPolicyCreated,
  TransferPolicyDestroyed: () => TransferPolicyDestroyed,
  TransferRequest: () => TransferRequest
});
module.exports = __toCommonJS(transfer_policy_exports);
var import_bcs = require("@mysten/sui/bcs");
var type_name = __toESM(require("../0x0000000000000000000000000000000000000000000000000000000000000001/type_name.js"));
var balance = __toESM(require("./balance.js"));
var object = __toESM(require("./object.js"));
var vec_set = __toESM(require("./vec_set.js"));
function TransferRequest() {
  return import_bcs.bcs.struct("TransferRequest", {
    item: import_bcs.bcs.Address,
    paid: import_bcs.bcs.u64(),
    from: import_bcs.bcs.Address,
    receipts: vec_set.VecSet(type_name.TypeName())
  });
}
function TransferPolicy() {
  return import_bcs.bcs.struct("TransferPolicy", {
    id: object.UID(),
    balance: balance.Balance(),
    rules: vec_set.VecSet(type_name.TypeName())
  });
}
function TransferPolicyCap() {
  return import_bcs.bcs.struct("TransferPolicyCap", {
    id: object.UID(),
    policy_id: import_bcs.bcs.Address
  });
}
function TransferPolicyCreated() {
  return import_bcs.bcs.struct("TransferPolicyCreated", {
    id: import_bcs.bcs.Address
  });
}
function TransferPolicyDestroyed() {
  return import_bcs.bcs.struct("TransferPolicyDestroyed", {
    id: import_bcs.bcs.Address
  });
}
function RuleKey() {
  return import_bcs.bcs.struct("RuleKey", {
    dummy_field: import_bcs.bcs.bool()
  });
}
