// @mysten/walrus@0.0.19/dist/cjs/contracts/deps/0x0000000000000000000000000000000000000000000000000000000000000002/tx_context.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var tx_context_exports = {};
__export(tx_context_exports, {
  TxContext: () => TxContext
});
module.exports = __toCommonJS(tx_context_exports);
var import_bcs = require("@mysten/sui/bcs");
function TxContext() {
  return import_bcs.bcs.struct("TxContext", {
    sender: import_bcs.bcs.Address,
    tx_hash: import_bcs.bcs.vector(import_bcs.bcs.u8()),
    epoch: import_bcs.bcs.u64(),
    epoch_timestamp_ms: import_bcs.bcs.u64(),
    ids_created: import_bcs.bcs.u64()
  });
}
