// @mysten/walrus@0.0.19/dist/cjs/contracts/deps/0x0000000000000000000000000000000000000000000000000000000000000002/zklogin_verified_id.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var zklogin_verified_id_exports = {};
__export(zklogin_verified_id_exports, {
  VerifiedID: () => VerifiedID
});
module.exports = __toCommonJS(zklogin_verified_id_exports);
var import_bcs = require("@mysten/sui/bcs");
var object = __toESM(require("./object.js"));
function VerifiedID() {
  return import_bcs.bcs.struct("VerifiedID", {
    id: object.UID(),
    owner: import_bcs.bcs.Address,
    key_claim_name: import_bcs.bcs.string(),
    key_claim_value: import_bcs.bcs.string(),
    issuer: import_bcs.bcs.string(),
    audience: import_bcs.bcs.string()
  });
}
