// @mysten/walrus@0.0.19/dist/esm/contracts/deps/0x0000000000000000000000000000000000000000000000000000000000000002/coin.js
import { bcs } from "@mysten/sui/bcs";
import * as balance from "./balance.js";
import * as object from "./object.js";
import * as url from "./url.js";
function Coin() {
  return bcs.struct("Coin", {
    id: object.UID(),
    balance: balance.Balance()
  });
}
function CoinMetadata() {
  return bcs.struct("CoinMetadata", {
    id: object.UID(),
    decimals: bcs.u8(),
    name: bcs.string(),
    symbol: bcs.string(),
    description: bcs.string(),
    icon_url: bcs.option(url.Url())
  });
}
function RegulatedCoinMetadata() {
  return bcs.struct("RegulatedCoinMetadata", {
    id: object.UID(),
    coin_metadata_object: bcs.Address,
    deny_cap_object: bcs.Address
  });
}
function TreasuryCap() {
  return bcs.struct("TreasuryCap", {
    id: object.UID(),
    total_supply: balance.Supply()
  });
}
function DenyCapV2() {
  return bcs.struct("DenyCapV2", {
    id: object.UID(),
    allow_global_pause: bcs.bool()
  });
}
function CurrencyCreated() {
  return bcs.struct("CurrencyCreated", {
    decimals: bcs.u8()
  });
}
function DenyCap() {
  return bcs.struct("DenyCap", {
    id: object.UID()
  });
}
export {
  Coin,
  CoinMetadata,
  CurrencyCreated,
  DenyCap,
  DenyCapV2,
  RegulatedCoinMetadata,
  TreasuryCap
};
