// @mysten/walrus@0.0.19/dist/esm/contracts/deps/0x0000000000000000000000000000000000000000000000000000000000000002/transfer_policy.js
import { bcs } from "@mysten/sui/bcs";
import * as type_name from "../0x0000000000000000000000000000000000000000000000000000000000000001/type_name.js";
import * as balance from "./balance.js";
import * as object from "./object.js";
import * as vec_set from "./vec_set.js";
function TransferRequest() {
  return bcs.struct("TransferRequest", {
    item: bcs.Address,
    paid: bcs.u64(),
    from: bcs.Address,
    receipts: vec_set.VecSet(type_name.TypeName())
  });
}
function TransferPolicy() {
  return bcs.struct("TransferPolicy", {
    id: object.UID(),
    balance: balance.Balance(),
    rules: vec_set.VecSet(type_name.TypeName())
  });
}
function TransferPolicyCap() {
  return bcs.struct("TransferPolicyCap", {
    id: object.UID(),
    policy_id: bcs.Address
  });
}
function TransferPolicyCreated() {
  return bcs.struct("TransferPolicyCreated", {
    id: bcs.Address
  });
}
function TransferPolicyDestroyed() {
  return bcs.struct("TransferPolicyDestroyed", {
    id: bcs.Address
  });
}
function RuleKey() {
  return bcs.struct("RuleKey", {
    dummy_field: bcs.bool()
  });
}
export {
  RuleKey,
  TransferPolicy,
  TransferPolicyCap,
  TransferPolicyCreated,
  TransferPolicyDestroyed,
  TransferRequest
};
