// @mysten/walrus@0.0.19/dist/cjs/contracts/apportionment_queue.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var apportionment_queue_exports = {};
__export(apportionment_queue_exports, {
  ApportionmentQueue: () => ApportionmentQueue,
  Entry: () => Entry,
  init: () => init
});
module.exports = __toCommonJS(apportionment_queue_exports);
var import_bcs = require("@mysten/sui/bcs");
var uq64_64 = __toESM(require("./deps/0x0000000000000000000000000000000000000000000000000000000000000001/uq64_64.js"));
var import_utils = require("./utils/index.js");
function ApportionmentQueue(...typeParameters) {
  return import_bcs.bcs.struct("ApportionmentQueue", {
    entries: import_bcs.bcs.vector(Entry(typeParameters[0]))
  });
}
function Entry(...typeParameters) {
  return import_bcs.bcs.struct("Entry", {
    priority: uq64_64.UQ64_64(),
    tie_breaker: import_bcs.bcs.u64(),
    value: typeParameters[0]
  });
}
function init(packageAddress) {
  function _new(options) {
    const argumentsTypes = [];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "apportionment_queue",
      function: "new",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes),
      typeArguments: options.typeArguments
    });
  }
  function pop_max(options) {
    const argumentsTypes = [
      `${packageAddress}::apportionment_queue::ApportionmentQueue<${options.typeArguments[0]}>`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "apportionment_queue",
      function: "pop_max",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes),
      typeArguments: options.typeArguments
    });
  }
  function insert(options) {
    const argumentsTypes = [
      `${packageAddress}::apportionment_queue::ApportionmentQueue<${options.typeArguments[0]}>`,
      "0x0000000000000000000000000000000000000000000000000000000000000001::uq64_64::UQ64_64",
      "u64",
      `${options.typeArguments[0]}`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "apportionment_queue",
      function: "insert",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes),
      typeArguments: options.typeArguments
    });
  }
  function bubble_down(options) {
    const argumentsTypes = [
      `vector<${packageAddress}::apportionment_queue::Entry<${options.typeArguments[0]}>>`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "apportionment_queue",
      function: "bubble_down",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes),
      typeArguments: options.typeArguments
    });
  }
  function bubble_up(options) {
    const argumentsTypes = [
      `vector<${packageAddress}::apportionment_queue::Entry<${options.typeArguments[0]}>>`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "apportionment_queue",
      function: "bubble_up",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes),
      typeArguments: options.typeArguments
    });
  }
  function higher_priority_than(options) {
    const argumentsTypes = [
      `${packageAddress}::apportionment_queue::Entry<${options.typeArguments[0]}>`,
      `${packageAddress}::apportionment_queue::Entry<${options.typeArguments[0]}>`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "apportionment_queue",
      function: "higher_priority_than",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes),
      typeArguments: options.typeArguments
    });
  }
  return { _new, pop_max, insert, bubble_down, bubble_up, higher_priority_than };
}
