// @mysten/walrus@0.0.19/dist/cjs/contracts/display.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var display_exports = {};
__export(display_exports, {
  ObjectDisplay: () => ObjectDisplay,
  PublisherKey: () => PublisherKey,
  init: () => init
});
module.exports = __toCommonJS(display_exports);
var import_bcs = require("@mysten/sui/bcs");
var object_bag = __toESM(require("./deps/0x0000000000000000000000000000000000000000000000000000000000000002/object_bag.js"));
var object = __toESM(require("./deps/0x0000000000000000000000000000000000000000000000000000000000000002/object.js"));
var import_utils = require("./utils/index.js");
function ObjectDisplay() {
  return import_bcs.bcs.struct("ObjectDisplay", {
    id: object.UID(),
    inner: object_bag.ObjectBag()
  });
}
function PublisherKey() {
  return import_bcs.bcs.struct("PublisherKey", {
    dummy_field: import_bcs.bcs.bool()
  });
}
function init(packageAddress) {
  function create(options) {
    const argumentsTypes = [
      "0x0000000000000000000000000000000000000000000000000000000000000002::package::Publisher"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "display",
      function: "create",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function init_blob_display(options) {
    const argumentsTypes = [
      "0x0000000000000000000000000000000000000000000000000000000000000002::package::Publisher"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "display",
      function: "init_blob_display",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function init_storage_display(options) {
    const argumentsTypes = [
      "0x0000000000000000000000000000000000000000000000000000000000000002::package::Publisher"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "display",
      function: "init_storage_display",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function init_staked_wal_display(options) {
    const argumentsTypes = [
      "0x0000000000000000000000000000000000000000000000000000000000000002::package::Publisher"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "display",
      function: "init_staked_wal_display",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  return { create, init_blob_display, init_storage_display, init_staked_wal_display };
}
