// @mysten/walrus@0.0.19/dist/cjs/contracts/events.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var events_exports = {};
__export(events_exports, {
  BlobCertified: () => BlobCertified,
  BlobDeleted: () => BlobDeleted,
  BlobRegistered: () => BlobRegistered,
  ContractUpgradeProposed: () => ContractUpgradeProposed,
  ContractUpgradeQuorumReached: () => ContractUpgradeQuorumReached,
  ContractUpgraded: () => ContractUpgraded,
  DenyListBlobDeleted: () => DenyListBlobDeleted,
  DenyListUpdate: () => DenyListUpdate,
  EpochChangeDone: () => EpochChangeDone,
  EpochChangeStart: () => EpochChangeStart,
  EpochParametersSelected: () => EpochParametersSelected,
  InvalidBlobID: () => InvalidBlobID,
  RegisterDenyListUpdate: () => RegisterDenyListUpdate,
  ShardRecoveryStart: () => ShardRecoveryStart,
  ShardsReceived: () => ShardsReceived,
  init: () => init
});
module.exports = __toCommonJS(events_exports);
var import_bcs = require("@mysten/sui/bcs");
var import_utils = require("./utils/index.js");
function BlobRegistered() {
  return import_bcs.bcs.struct("BlobRegistered", {
    epoch: import_bcs.bcs.u32(),
    blob_id: import_bcs.bcs.u256(),
    size: import_bcs.bcs.u64(),
    encoding_type: import_bcs.bcs.u8(),
    end_epoch: import_bcs.bcs.u32(),
    deletable: import_bcs.bcs.bool(),
    object_id: import_bcs.bcs.Address
  });
}
function BlobCertified() {
  return import_bcs.bcs.struct("BlobCertified", {
    epoch: import_bcs.bcs.u32(),
    blob_id: import_bcs.bcs.u256(),
    end_epoch: import_bcs.bcs.u32(),
    deletable: import_bcs.bcs.bool(),
    object_id: import_bcs.bcs.Address,
    is_extension: import_bcs.bcs.bool()
  });
}
function BlobDeleted() {
  return import_bcs.bcs.struct("BlobDeleted", {
    epoch: import_bcs.bcs.u32(),
    blob_id: import_bcs.bcs.u256(),
    end_epoch: import_bcs.bcs.u32(),
    object_id: import_bcs.bcs.Address,
    was_certified: import_bcs.bcs.bool()
  });
}
function InvalidBlobID() {
  return import_bcs.bcs.struct("InvalidBlobID", {
    epoch: import_bcs.bcs.u32(),
    blob_id: import_bcs.bcs.u256()
  });
}
function EpochChangeStart() {
  return import_bcs.bcs.struct("EpochChangeStart", {
    epoch: import_bcs.bcs.u32()
  });
}
function EpochChangeDone() {
  return import_bcs.bcs.struct("EpochChangeDone", {
    epoch: import_bcs.bcs.u32()
  });
}
function ShardsReceived() {
  return import_bcs.bcs.struct("ShardsReceived", {
    epoch: import_bcs.bcs.u32(),
    shards: import_bcs.bcs.vector(import_bcs.bcs.u16())
  });
}
function EpochParametersSelected() {
  return import_bcs.bcs.struct("EpochParametersSelected", {
    next_epoch: import_bcs.bcs.u32()
  });
}
function ShardRecoveryStart() {
  return import_bcs.bcs.struct("ShardRecoveryStart", {
    epoch: import_bcs.bcs.u32(),
    shards: import_bcs.bcs.vector(import_bcs.bcs.u16())
  });
}
function ContractUpgraded() {
  return import_bcs.bcs.struct("ContractUpgraded", {
    epoch: import_bcs.bcs.u32(),
    package_id: import_bcs.bcs.Address,
    version: import_bcs.bcs.u64()
  });
}
function RegisterDenyListUpdate() {
  return import_bcs.bcs.struct("RegisterDenyListUpdate", {
    epoch: import_bcs.bcs.u32(),
    root: import_bcs.bcs.u256(),
    sequence_number: import_bcs.bcs.u64(),
    node_id: import_bcs.bcs.Address
  });
}
function DenyListUpdate() {
  return import_bcs.bcs.struct("DenyListUpdate", {
    epoch: import_bcs.bcs.u32(),
    root: import_bcs.bcs.u256(),
    sequence_number: import_bcs.bcs.u64(),
    node_id: import_bcs.bcs.Address
  });
}
function DenyListBlobDeleted() {
  return import_bcs.bcs.struct("DenyListBlobDeleted", {
    epoch: import_bcs.bcs.u32(),
    blob_id: import_bcs.bcs.u256()
  });
}
function ContractUpgradeProposed() {
  return import_bcs.bcs.struct("ContractUpgradeProposed", {
    epoch: import_bcs.bcs.u32(),
    package_digest: import_bcs.bcs.vector(import_bcs.bcs.u8())
  });
}
function ContractUpgradeQuorumReached() {
  return import_bcs.bcs.struct("ContractUpgradeQuorumReached", {
    epoch: import_bcs.bcs.u32(),
    package_digest: import_bcs.bcs.vector(import_bcs.bcs.u8())
  });
}
function init(packageAddress) {
  function emit_blob_registered(options) {
    const argumentsTypes = [
      "u32",
      "u256",
      "u64",
      "u8",
      "u32",
      "bool",
      "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_blob_registered",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function emit_blob_certified(options) {
    const argumentsTypes = [
      "u32",
      "u256",
      "u32",
      "bool",
      "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
      "bool"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_blob_certified",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function emit_invalid_blob_id(options) {
    const argumentsTypes = ["u32", "u256"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_invalid_blob_id",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function emit_blob_deleted(options) {
    const argumentsTypes = [
      "u32",
      "u256",
      "u32",
      "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
      "bool"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_blob_deleted",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function emit_epoch_change_start(options) {
    const argumentsTypes = ["u32"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_epoch_change_start",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function emit_epoch_change_done(options) {
    const argumentsTypes = ["u32"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_epoch_change_done",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function emit_shards_received(options) {
    const argumentsTypes = ["u32", "vector<u16>"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_shards_received",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function emit_epoch_parameters_selected(options) {
    const argumentsTypes = ["u32"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_epoch_parameters_selected",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function emit_shard_recovery_start(options) {
    const argumentsTypes = ["u32", "vector<u16>"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_shard_recovery_start",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function emit_contract_upgraded(options) {
    const argumentsTypes = [
      "u32",
      "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID",
      "u64"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_contract_upgraded",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function emit_register_deny_list_update(options) {
    const argumentsTypes = [
      "u32",
      "u256",
      "u64",
      "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_register_deny_list_update",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function emit_deny_list_update(options) {
    const argumentsTypes = [
      "u32",
      "u256",
      "u64",
      "0x0000000000000000000000000000000000000000000000000000000000000002::object::ID"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_deny_list_update",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function emit_deny_listed_blob_deleted(options) {
    const argumentsTypes = ["u32", "u256"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_deny_listed_blob_deleted",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function emit_contract_upgrade_proposed(options) {
    const argumentsTypes = ["u32", "vector<u8>"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_contract_upgrade_proposed",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function emit_contract_upgrade_quorum_reached(options) {
    const argumentsTypes = ["u32", "vector<u8>"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "events",
      function: "emit_contract_upgrade_quorum_reached",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  return {
    emit_blob_registered,
    emit_blob_certified,
    emit_invalid_blob_id,
    emit_blob_deleted,
    emit_epoch_change_start,
    emit_epoch_change_done,
    emit_shards_received,
    emit_epoch_parameters_selected,
    emit_shard_recovery_start,
    emit_contract_upgraded,
    emit_register_deny_list_update,
    emit_deny_list_update,
    emit_deny_listed_blob_deleted,
    emit_contract_upgrade_proposed,
    emit_contract_upgrade_quorum_reached
  };
}
