// @mysten/walrus@0.0.19/dist/cjs/contracts/extended_field.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var extended_field_exports = {};
__export(extended_field_exports, {
  ExtendedField: () => ExtendedField,
  Key: () => Key,
  init: () => init
});
module.exports = __toCommonJS(extended_field_exports);
var import_bcs = require("@mysten/sui/bcs");
var object = __toESM(require("./deps/0x0000000000000000000000000000000000000000000000000000000000000002/object.js"));
var import_utils = require("./utils/index.js");
function ExtendedField() {
  return import_bcs.bcs.struct("ExtendedField", {
    id: object.UID()
  });
}
function Key() {
  return import_bcs.bcs.struct("Key", {
    dummy_field: import_bcs.bcs.bool()
  });
}
function init(packageAddress) {
  function _new(options) {
    const argumentsTypes = [`${options.typeArguments[0]}`];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "extended_field",
      function: "new",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes),
      typeArguments: options.typeArguments
    });
  }
  function borrow(options) {
    const argumentsTypes = [
      `${packageAddress}::extended_field::ExtendedField<${options.typeArguments[0]}>`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "extended_field",
      function: "borrow",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes),
      typeArguments: options.typeArguments
    });
  }
  function borrow_mut(options) {
    const argumentsTypes = [
      `${packageAddress}::extended_field::ExtendedField<${options.typeArguments[0]}>`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "extended_field",
      function: "borrow_mut",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes),
      typeArguments: options.typeArguments
    });
  }
  function swap(options) {
    const argumentsTypes = [
      `${packageAddress}::extended_field::ExtendedField<${options.typeArguments[0]}>`,
      `${options.typeArguments[0]}`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "extended_field",
      function: "swap",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes),
      typeArguments: options.typeArguments
    });
  }
  function destroy(options) {
    const argumentsTypes = [
      `${packageAddress}::extended_field::ExtendedField<${options.typeArguments[0]}>`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "extended_field",
      function: "destroy",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes),
      typeArguments: options.typeArguments
    });
  }
  return { _new, borrow, borrow_mut, swap, destroy };
}
