// @mysten/walrus@0.0.19/dist/cjs/contracts/pool_exchange_rate.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var pool_exchange_rate_exports = {};
__export(pool_exchange_rate_exports, {
  PoolExchangeRate: () => PoolExchangeRate,
  init: () => init
});
module.exports = __toCommonJS(pool_exchange_rate_exports);
var import_bcs = require("@mysten/sui/bcs");
var import_utils = require("./utils/index.js");
function PoolExchangeRate() {
  return import_bcs.bcs.enum("PoolExchangeRate", {
    Flat: null,
    Variable: import_bcs.bcs.tuple([import_bcs.bcs.u128(), import_bcs.bcs.u128()])
  });
}
function init(packageAddress) {
  function flat(options) {
    const argumentsTypes = [];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "pool_exchange_rate",
      function: "flat",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function _new(options) {
    const argumentsTypes = ["u64", "u64"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "pool_exchange_rate",
      function: "new",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function convert_to_wal_amount(options) {
    const argumentsTypes = [`${packageAddress}::pool_exchange_rate::PoolExchangeRate`, "u64"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "pool_exchange_rate",
      function: "convert_to_wal_amount",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function convert_to_share_amount(options) {
    const argumentsTypes = [`${packageAddress}::pool_exchange_rate::PoolExchangeRate`, "u64"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "pool_exchange_rate",
      function: "convert_to_share_amount",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  return { flat, _new, convert_to_wal_amount, convert_to_share_amount };
}
