// @mysten/walrus@0.0.19/dist/cjs/contracts/storage_resource.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var storage_resource_exports = {};
__export(storage_resource_exports, {
  Storage: () => Storage,
  init: () => init
});
module.exports = __toCommonJS(storage_resource_exports);
var import_bcs = require("@mysten/sui/bcs");
var object = __toESM(require("./deps/0x0000000000000000000000000000000000000000000000000000000000000002/object.js"));
var import_utils = require("./utils/index.js");
function Storage() {
  return import_bcs.bcs.struct("Storage", {
    id: object.UID(),
    start_epoch: import_bcs.bcs.u32(),
    end_epoch: import_bcs.bcs.u32(),
    storage_size: import_bcs.bcs.u64()
  });
}
function init(packageAddress) {
  function start_epoch(options) {
    const argumentsTypes = [`${packageAddress}::storage_resource::Storage`];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "storage_resource",
      function: "start_epoch",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function end_epoch(options) {
    const argumentsTypes = [`${packageAddress}::storage_resource::Storage`];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "storage_resource",
      function: "end_epoch",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function size(options) {
    const argumentsTypes = [`${packageAddress}::storage_resource::Storage`];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "storage_resource",
      function: "size",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function create_storage(options) {
    const argumentsTypes = ["u32", "u32", "u64"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "storage_resource",
      function: "create_storage",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function extend_end_epoch(options) {
    const argumentsTypes = [`${packageAddress}::storage_resource::Storage`, "u32"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "storage_resource",
      function: "extend_end_epoch",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function split_by_epoch(options) {
    const argumentsTypes = [`${packageAddress}::storage_resource::Storage`, "u32"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "storage_resource",
      function: "split_by_epoch",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function split_by_size(options) {
    const argumentsTypes = [`${packageAddress}::storage_resource::Storage`, "u64"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "storage_resource",
      function: "split_by_size",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function fuse_periods(options) {
    const argumentsTypes = [
      `${packageAddress}::storage_resource::Storage`,
      `${packageAddress}::storage_resource::Storage`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "storage_resource",
      function: "fuse_periods",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function fuse_amount(options) {
    const argumentsTypes = [
      `${packageAddress}::storage_resource::Storage`,
      `${packageAddress}::storage_resource::Storage`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "storage_resource",
      function: "fuse_amount",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function fuse(options) {
    const argumentsTypes = [
      `${packageAddress}::storage_resource::Storage`,
      `${packageAddress}::storage_resource::Storage`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "storage_resource",
      function: "fuse",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function destroy(options) {
    const argumentsTypes = [`${packageAddress}::storage_resource::Storage`];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "storage_resource",
      function: "destroy",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  return {
    start_epoch,
    end_epoch,
    size,
    create_storage,
    extend_end_epoch,
    split_by_epoch,
    split_by_size,
    fuse_periods,
    fuse_amount,
    fuse,
    destroy
  };
}
