// @mysten/walrus@0.0.19/dist/cjs/contracts/system_state_inner.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var system_state_inner_exports = {};
__export(system_state_inner_exports, {
  SystemStateInnerV1: () => SystemStateInnerV1,
  init: () => init
});
module.exports = __toCommonJS(system_state_inner_exports);
var import_bcs = require("@mysten/sui/bcs");
var bls_aggregate = __toESM(require("./bls_aggregate.js"));
var event_blob = __toESM(require("./event_blob.js"));
var extended_field = __toESM(require("./extended_field.js"));
var storage_accounting = __toESM(require("./storage_accounting.js"));
var import_utils = require("./utils/index.js");
function SystemStateInnerV1() {
  return import_bcs.bcs.struct("SystemStateInnerV1", {
    committee: bls_aggregate.BlsCommittee(),
    total_capacity_size: import_bcs.bcs.u64(),
    used_capacity_size: import_bcs.bcs.u64(),
    storage_price_per_unit_size: import_bcs.bcs.u64(),
    write_price_per_unit_size: import_bcs.bcs.u64(),
    future_accounting: storage_accounting.FutureAccountingRingBuffer(),
    event_blob_certification_state: event_blob.EventBlobCertificationState(),
    deny_list_sizes: extended_field.ExtendedField()
  });
}
function init(packageAddress) {
  function create_empty(options) {
    const argumentsTypes = ["u32"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "create_empty",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function advance_epoch(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      `${packageAddress}::bls_aggregate::BlsCommittee`,
      `${packageAddress}::epoch_parameters::EpochParams`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "advance_epoch",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function reserve_space(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      "u64",
      "u32",
      `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::wal::WAL>`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "reserve_space",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function reserve_space_without_payment(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      "u64",
      "u32",
      "bool"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "reserve_space_without_payment",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function invalidate_blob_id(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      "vector<u8>",
      "vector<u8>",
      "vector<u8>"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "invalidate_blob_id",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function register_blob(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      `${packageAddress}::storage_resource::Storage`,
      "u256",
      "u256",
      "u64",
      "u8",
      "bool",
      `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::wal::WAL>`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "register_blob",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function certify_blob(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      `${packageAddress}::blob::Blob`,
      "vector<u8>",
      "vector<u8>",
      "vector<u8>"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "certify_blob",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function delete_blob(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      `${packageAddress}::blob::Blob`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "delete_blob",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function extend_blob_with_resource(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      `${packageAddress}::blob::Blob`,
      `${packageAddress}::storage_resource::Storage`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "extend_blob_with_resource",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function extend_blob(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      `${packageAddress}::blob::Blob`,
      "u32",
      `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::wal::WAL>`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "extend_blob",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function process_storage_payments(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      "u64",
      "u32",
      "u32",
      `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::wal::WAL>`
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "process_storage_payments",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function certify_event_blob(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      `${packageAddress}::storage_node::StorageNodeCap`,
      "u256",
      "u256",
      "u64",
      "u8",
      "u64",
      "u32"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "certify_event_blob",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function add_subsidy(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::wal::WAL>`,
      "u32"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "add_subsidy",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function epoch(options) {
    const argumentsTypes = [`${packageAddress}::system_state_inner::SystemStateInnerV1`];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "epoch",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function total_capacity_size(options) {
    const argumentsTypes = [`${packageAddress}::system_state_inner::SystemStateInnerV1`];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "total_capacity_size",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function used_capacity_size(options) {
    const argumentsTypes = [`${packageAddress}::system_state_inner::SystemStateInnerV1`];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "used_capacity_size",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function committee(options) {
    const argumentsTypes = [`${packageAddress}::system_state_inner::SystemStateInnerV1`];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "committee",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function n_shards(options) {
    const argumentsTypes = [`${packageAddress}::system_state_inner::SystemStateInnerV1`];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "n_shards",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function write_price(options) {
    const argumentsTypes = [`${packageAddress}::system_state_inner::SystemStateInnerV1`, "u64"];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "write_price",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function register_deny_list_update(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      `${packageAddress}::storage_node::StorageNodeCap`,
      "u256",
      "u64"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "register_deny_list_update",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function update_deny_list(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      `${packageAddress}::storage_node::StorageNodeCap`,
      "vector<u8>",
      "vector<u8>",
      "vector<u8>"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "update_deny_list",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function delete_deny_listed_blob(options) {
    const argumentsTypes = [
      `${packageAddress}::system_state_inner::SystemStateInnerV1`,
      "vector<u8>",
      "vector<u8>",
      "vector<u8>"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "system_state_inner",
      function: "delete_deny_listed_blob",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  return {
    create_empty,
    advance_epoch,
    reserve_space,
    reserve_space_without_payment,
    invalidate_blob_id,
    register_blob,
    certify_blob,
    delete_blob,
    extend_blob_with_resource,
    extend_blob,
    process_storage_payments,
    certify_event_blob,
    add_subsidy,
    epoch,
    total_capacity_size,
    used_capacity_size,
    committee,
    n_shards,
    write_price,
    register_deny_list_update,
    update_deny_list,
    delete_deny_listed_blob
  };
}
