// @mysten/walrus@0.0.19/dist/cjs/contracts/walrus_context.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var walrus_context_exports = {};
__export(walrus_context_exports, {
  WalrusContext: () => WalrusContext,
  init: () => init
});
module.exports = __toCommonJS(walrus_context_exports);
var import_bcs = require("@mysten/sui/bcs");
var vec_map = __toESM(require("./deps/0x0000000000000000000000000000000000000000000000000000000000000002/vec_map.js"));
var import_utils = require("./utils/index.js");
function WalrusContext() {
  return import_bcs.bcs.struct("WalrusContext", {
    epoch: import_bcs.bcs.u32(),
    committee_selected: import_bcs.bcs.bool(),
    committee: vec_map.VecMap(import_bcs.bcs.Address, import_bcs.bcs.vector(import_bcs.bcs.u16()))
  });
}
function init(packageAddress) {
  function _new(options) {
    const argumentsTypes = [
      "u32",
      "bool",
      "0x0000000000000000000000000000000000000000000000000000000000000002::vec_map::VecMap<0x0000000000000000000000000000000000000000000000000000000000000002::object::ID, vector<u16>>"
    ];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "walrus_context",
      function: "new",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function epoch(options) {
    const argumentsTypes = [`${packageAddress}::walrus_context::WalrusContext`];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "walrus_context",
      function: "epoch",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function committee_selected(options) {
    const argumentsTypes = [`${packageAddress}::walrus_context::WalrusContext`];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "walrus_context",
      function: "committee_selected",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  function committee(options) {
    const argumentsTypes = [`${packageAddress}::walrus_context::WalrusContext`];
    return (tx) => tx.moveCall({
      package: packageAddress,
      module: "walrus_context",
      function: "committee",
      arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes)
    });
  }
  return { _new, epoch, committee_selected, committee };
}
